accepted/2021/package-validation/apicompat.md

```diff
@@ -0,0 +1,143 @@
+3. Return types match.
+
+### 10. Modifiers cannot change
+This rule checks for differences in marshaling attributes like `in`, `out`, `ref`, `readonly` and custom modifiers like `const` and `volatile` in parameters and member return's type.
+
+1. Return's type modifiers can be ref, readonly or none. 
+2. Parameter modifiers can be in, out, ref and custom modifiers like const and volatile can exist.
+
+### 11. Parameter names cannot change
+This rule makes sure that constructor, property or method parameter names match.
+
+### 12. Type cannot change classification
+This rule makes sure that type's classification doesn't change. Type classification refers to, `class`, `struct`, `ref struct`, `readonly`, `interface` or `delegate`.
+
+### 13. Nullable annotations should match
+This rule should make sure that nullable annotations match in between the implementation and the contract.
+
+Also, this rule should allow differences in nullable annotations that are not breaking, for example, relaxing a nullable reference type annotation on an in parameter (i.e `public string Foo(string a) -> public string Foo(string? a)`). However, this should be configurable via rule settings to disable this as in some cases where you are comparing a reference assembly vs it's implementation you want them to match in all cases.
+
+### 14. Constructor makes base class instantiable.
+Rule that checks if a public API is a constructor that makes a class instantiable, even though the base class is not instantiable. That API pattern is not allowed, because it causes protected members of the base class, which are not considered public APIs, to be exposed to subclasses of this class.
+
+## Configurability
+
+Given that we are proposing a Roslyn analyzer a dotnet tool and potentially an MSBuild Task, we should support a friendly format for all three possible scenarios to baseline warnings/errors, configure rules and severity of rules.
+
+To baseline errors we could support:
+
+1. All analyzer's disable mechanism when using the tool as an analyzer.
+2. A txt file that lists the errors to baseline (maybe we could support this for the analyzer scenario as well).
+
+Configuring rules involves settings like, ignoring certain attributes as describe in the attribute compat rule, rules number 8 in the document, etc. I think these could be part of the `.editorconfig` with configuration for rules' severity.
+
+### MSBuild support
+
+We should provide MSBuild first class configurability via properties so that costumers can just add a `PackageReference` and set properties on how they want to consume the tool (via MSBuild Task, Analyzer, etc), what the inputs are going to be, where the baseline errors file can be found, etc. This would make it very simple for customers to include in their libraries and configure without having to write complicated MSBuild targets to run the tool on an MSBuild process if that is their best way to hook it up into their build system.
```

Comment:
> Roslyn analyzer a dotnet tool and potentially an MSBuild Task,

Could you say a little more about the scenarios for each? I guess Roslyn analyzers don't get handed the list of references, or we wouldn't need an MSBuild task also? When would we use each of these -- and do we need all three in this release?

Reply:
Maybe this goes together with your list of different inputs.

Reply:
Not to belabor the point, -- but I think it would be helpful to walk through from a customer point of view -- they have an existing library they're building and shipping, and want to be sure they don't unintentionally break the surface -- they heard about this feature -- what do they do now? Other scenarios might include -- understanding whether their dependencies changed ? Baselining a known break? Baselines that differ according to target framework? Etc. There may be other scenarios.

Reply:
Sure, makes sense. I will also add priority of each one, user scenarios, pros and cons.